### PR TITLE
Cleanup helper functions to validate search results

### PR DESCRIPTION
**Pull request by dalg24**
Please do not get hung up in the details of the helper function to pack/unpack tuples and manipulate compressed sparse row format.

This PR is meant to simplify and unify the comparison of the search with a reference solution.

### CODE REVIEW

File: test/CompressedSparseRow.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+#include "VectorOfTuples.hpp"
+
+namespace ext
+{
+
+// Extension for compile-time sequence of integers that defines a helper class
+// template make_index_sequence_with_offset to simplify the creation of
+// std::index_sequence type with O, O+1, O+2, O+N-1 as Ints.
+template <std::size_t O, std::size_t... Ints>
+constexpr std::index_sequence<(O + Ints)...>
```
Comment (masterleinad) [inline test/CompressedSparseRow.hpp]: Any reason for choosing `O` here? I guess you want this to look similar to 0?

Comment (dalg24) [reply test/CompressedSparseRow.hpp]: I realize now it confusing.  I can change if you like.

**Comment by masterleinad** [comment]
```
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(24): error: namespace "std" has no member "index_sequence"
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(24): error: "constexpr" is not valid here
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(34): warning: parsing restarts here after previous syntax error
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(39): error: namespace "std" has no member "index_sequence"
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(39): error: expected a ")"
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(39): error: "auto" function requires a trailing return type
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(45): error: "auto" function requires a trailing return type
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(66): error: "auto" function requires a trailing return type
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(75): error: namespace "std" has no member "index_sequence"
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(75): error: expected a ")"
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(74): error: "auto" function requires a trailing return type
/var/jenkins/workspace/ArborX_PR-81/test/CompressedSparseRow.hpp(83): error: "auto" function requires a trailing return type
```

File: test/CompressedSparseRow.hpp (hunk 1 above)
Comment (masterleinad) [reply test/CompressedSparseRow.hpp]: Thanks!

**Comment by dalg24** [comment]
@masterleinad that was the point of #82 
I will rebase

File: test/CompressedSparseRow.hpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#define ARBORX_COMPRESSED_SPARSE_ROW_HPP
+
+#include "VectorOfTuples.hpp"
+
+namespace ext
+{
+
+// Extension for compile-time sequence of integers that defines a helper class
+// template make_index_sequence_with_offset to simplify the creation of
+// std::index_sequence type with O, O+1, O+2, O+N-1 as Ints.
```
Comment (masterleinad) [inline test/CompressedSparseRow.hpp]: Can you please use the same name in the comments as in the code?

File: test/Search_UnitTestHelpers.hpp [hunk 3]
```diff
@@ -27,6 +33,48 @@
...
+  }
+};
+
+} // namespace Details
+
+template <typename T1, typename T2>
+void validateResults(T1 const &reference, T2 const &other)
+{
+  auto const m = getNumberOfRows(reference);
+  BOOST_TEST(m == getNumberOfRows(reference));
```
Comment (dalg24) [inline test/Search_UnitTestHelpers.hpp]: oops

File: test/Search_UnitTestHelpers.hpp [hunk 4]
```diff
@@ -39,13 +87,13 @@ void checkResults(ArborX::BVH<DeviceType> const &bvh,
   Kokkos::View<int *, DeviceType> offset("offset", 0);
   bvh.query(queries, indices, offset);
 
-  auto indices_host = Kokkos::create_mirror_view(indices);
-  deep_copy(indices_host, indices);
-  auto offset_host = Kokkos::create_mirror_view(offset);
-  deep_copy(offset_host, offset);
+  auto indices_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, indices);
```
Comment (masterleinad) [inline test/Search_UnitTestHelpers.hpp]: Nice!

File: test/VectorOfTuples.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  for (std::size_t i = 0; i < n; ++i)
+  {
+    out.emplace_back(Details::ArrayTraits<Arrays>::access(in, i)...);
+  }
+  return out;
+}
+
+template <typename... Arrays>
+std::vector<std::tuple<typename Details::ArrayTraits<Arrays>::value_type...>>
+subsetToVectorOfTuples(std::size_t begin, std::size_t end, Arrays const &... in)
```
Comment (dalg24) [inline test/VectorOfTuples.hpp]: This function was introduced in this PR and I did not bother testing (nor plan to)

File: test/CompressedSparseRow.hpp (hunk 2 above)
Comment (dalg24) [reply test/CompressedSparseRow.hpp]: Done

File: test/VectorOfTuples.hpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+} // namespace Details
+
+template <typename... Arrays>
+std::vector<std::tuple<typename Details::ArrayTraits<Arrays>::value_type...>>
+toVectorOfTuples(Arrays const &... in)
+{
+  std::vector<std::tuple<typename Details::ArrayTraits<Arrays>::value_type...>>
+      out;
+  std::size_t const n = Details::getSizeOfArrays(in...);
+  for (std::size_t i = 0; i < n; ++i)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply test/VectorOfTuples.hpp]: With your suggestion to assert that `end <= getSizeOfArrays(in...)` in the other function, I think that actually makes sense to delegate to that function.  Will change.

File: test/CompressedSparseRow.hpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+} // namespace ext
+
+template <typename Tuple, std::size_t... Is>
+auto getSizeOfArraysInTuple(Tuple const &t, std::index_sequence<Is...>)
+{
+  return Details::getSizeOfArrays(std::get<Is>(t)...);
+}
+
+template <typename TupleOfArrays>
+auto getNNZ(TupleOfArrays const &t)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply test/CompressedSparseRow.hpp]: I was planning on using it to check the tuple for sanity but forgot to call it from the main routine.

File: test/CompressedSparseRow.hpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+{
+  using ArrayType = typename std::tuple_element<0, TupleOfArrays>::type;
+  using Traits = Details::ArrayTraits<typename std::remove_cv<ArrayType>::type>;
+  constexpr std::size_t N = std::tuple_size<TupleOfArrays>::value;
+  auto const nnz = getSizeOfArraysInTuple(
+      t, ext::make_index_sequence_with_offset<1, N - 1>{});
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply test/CompressedSparseRow.hpp]: Right. Forgot to remove that comment.

File: test/CompressedSparseRow.hpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    std::stringstream msg;
+    msg << "mismatch between number of entries in trailing arrays and last "
+           "element of first array"
+        << " (" << nnz << " != " << last_element_of_first_array << ')';
+    throw std::invalid_argument(msg.str());
+  }
+  return nnz;
+}
+
+template <typename TupleOfArrays>
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply test/CompressedSparseRow.hpp]: Yes, that makes more sense now.

**Comment by dalg24** [comment]
@masterleinad I believe I addressed most of your review comments.

File: test/CompressedSparseRow.hpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+auto getNNZ(TupleOfArrays const &t)
+{
+  using ArrayType = typename std::tuple_element<0, TupleOfArrays>::type;
+  using Traits = Details::ArrayTraits<typename std::remove_cv<ArrayType>::type>;
+  constexpr std::size_t N = std::tuple_size<TupleOfArrays>::value;
+  auto const nnz = getSizeOfArraysInTuple(
+      t, ext::make_index_sequence_with_offset<1, N - 1>{});
+  auto const last_element_of_first_array = static_cast<std::size_t>(
+      Traits::access(std::get<0>(t), Traits::size(std::get<0>(t)) - 1));
```
[the remark this answers is not in the document]
Comment (dalg24) [reply test/CompressedSparseRow.hpp]: Now I realize you probably meant `Traits::access(std::get<0>(t), getNumberOfRows(t))`.
Meh.  Won't fix.

File: test/CompressedSparseRow.hpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+template <typename Tuple, std::size_t... Is>
+auto subsetAndSort(Tuple const &t, std::size_t offset, std::size_t count,
+                   std::index_sequence<Is...>)
+{
+  auto out = subsetToVectorOfTuples(offset, count, std::get<Is>(t)...);
+  std::sort(std::begin(out), std::end(out));
+  return out;
+}
+
+template <typename TupleOfArrays>
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply test/CompressedSparseRow.hpp]: We are documenting where we expect special data now. That's OK.

File: test/Search_UnitTestHelpers.hpp [hunk 12]
```diff
@@ -27,6 +33,48 @@
...
+
+template <typename T, typename... Ps>
+struct ArrayTraits<Kokkos::View<T *, Ps...>>
+{
+  using array_type = Kokkos::View<T *, Ps...>;
+  using value_type = T;
+  static std::size_t size(array_type const &v) { return v.extent(0); }
+  static value_type const &access(array_type const &v, std::size_t i)
+  {
+    return v(i);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply test/Search_UnitTestHelpers.hpp]: Not that important anyway.

**Comment by masterleinad** [comment]
Would you mind documenting the remaining functions in `test/VectorOfTuples.hpp` as well (especially, `checkProperlySized`)? Feel free to merge after that.

**Pull request merged**